File: src/terceraParteEjerciciosTraduccion/ejercicio5/ejercicio5.java
package terceraParteEjerciciosTraduccion.ejercicio5;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Tarea {
    String descripcion;
    boolean completada;

    public Tarea(String descripcion) {
        this.descripcion = descripcion;
        this.completada = false;
    }

    public void marcarCompletada() {
        this.completada = true;
    }

    public String getEstado() {
        return completada ? "Completada" : "Pendiente";
    }

    @Override
    public String toString() {
        return descripcion + " - " + getEstado();
    }
}

class Proyecto {
    String nombre;
    List<Tarea> tareas;

    public Proyecto(String nombre) {
        this.nombre = nombre;
        this.tareas = new ArrayList<>();
    }

    public void agregarTarea(Tarea tarea) {
        tareas.add(tarea);
        System.out.println("Tarea \"" + tarea.descripcion + "\" añadida al proyecto " + nombre);
    }

    public void mostrarTareas() {
        if (tareas.isEmpty()) {
            System.out.println("No hay tareas en este proyecto.");
        } else {
            System.out.println("Tareas en el proyecto " + nombre + ":");
            for (int i = 0; i < tareas.size(); i++) {
                System.out.println((i + 1) + ". " + tareas.get(i));
            }
        }
    }

    public void marcarTareaCompletada(int indice) {
        if (indice >= 0 && indice < tareas.size()) {
            tareas.get(indice).marcarCompletada();
            System.out.println("Tarea marcada como completada.");
        } else {
            System.out.println("Índice de tarea no válido.");
        }
    }
}

class Usuario {
    String nombre;
    List<Proyecto> proyectos;

    public Usuario(String nombre) {
        this.nombre = nombre;
        this.proyectos = new ArrayList<>();
    }

    public void crearProyecto(String nombreProyecto) {
        Proyecto proyecto = new Proyecto(nombreProyecto);
        proyectos.add(proyecto);
        System.out.println("Proyecto \"" + nombreProyecto + "\" creado.");
    }

    public Proyecto obtenerProyecto(int indice) {
        if (indice >= 0 && indice < proyectos.size()) {
            return proyectos.get(indice);
        } else {
            System.out.println("Índice de proyecto no válido.");
            return null;
        }
    }

    public void mostrarProyectos() {
        if (proyectos.isEmpty()) {
            System.out.println("No hay proyectos disponibles.");
        } else {
            System.out.println("Proyectos de " + nombre + ":");
            for (int i = 0; i < proyectos.size(); i++) {
                System.out.println((i + 1) + ". " + proyectos.get(i).nombre);
            }
        }
    }
}

 class GestorDeTareas {
    private static final Scanner scanner = new Scanner(System.in);
    private static final Usuario usuario = new Usuario("Usuario1");

    public static void main(String[] args) {
        while (true) {
            System.out.println("\nGestor de Tareas");
            System.out.println("1. Crear proyecto");
            System.out.println("2. Agregar tarea a proyecto");
            System.out.println("3. Marcar tarea como completada");
            System.out.println("4. Mostrar proyectos y tareas");
            System.out.println("5. Salir");
            System.out.print("Seleccione una opción: ");
            int opcion = scanner.nextInt();
            scanner.nextLine();  // Limpiar el buffer

            switch (opcion) {
                case 1 -> crearProyecto();
                case 2 -> agregarTarea();
                case 3 -> marcarTareaCompletada();
                case 4 -> mostrarProyectosYTareas();
                case 5 -> {
                    System.out.println("Saliendo...");
                    scanner.close();
                    return;
                }
                default -> System.out.println("Opción no válida.");
            }
        }
    }

    private static void crearProyecto() {
        System.out.print("Ingrese el nombre del proyecto: ");
        String nombreProyecto = scanner.nextLine();
        usuario.crearProyecto(nombreProyecto);
    }

    private static void agregarTarea() {
        usuario.mostrarProyectos();
        System.out.print("Seleccione el número del proyecto: ");
        int indiceProyecto = scanner.nextInt() - 1;
        scanner.nextLine();  // Limpiar el buffer
        Proyecto proyecto = usuario.obtenerProyecto(indiceProyecto);

        if (proyecto != null) {
            System.out.print("Descripción de la tarea: ");
            String descripcionTarea = scanner.nextLine();
            proyecto.agregarTarea(new Tarea(descripcionTarea));
        }
    }

    private static void marcarTareaCompletada() {
        usuario.mostrarProyectos();
        System.out.print("Seleccione el número del proyecto: ");
        int indiceProyecto = scanner.nextInt() - 1;
        scanner.nextLine();  // Limpiar el buffer
        Proyecto proyecto = usuario.obtenerProyecto(indiceProyecto);

        if (proyecto != null) {
            proyecto.mostrarTareas();
            System.out.print("Seleccione el número de la tarea a completar: ");
            int indiceTarea = scanner.nextInt() - 1;
            scanner.nextLine();  // Limpiar el buffer
            proyecto.marcarTareaCompletada(indiceTarea);
        }
    }

    private static void mostrarProyectosYTareas() {
        usuario.mostrarProyectos();
        for (Proyecto proyecto : usuario.proyectos) {
            proyecto.mostrarTareas();
        }
    }
}

File: src/terceraParteEjerciciosTraduccion/ejercicio5/ejercicio5.kt
package terceraParteEjerciciosTraduccion.ejercicio5

// 5. Gestor de Tareas
// Diseña un gestor de tareas. Crea clases para Tarea, Proyecto, y Usuario. Permite a
// los usuarios crear proyectos, añadir tareas y marcar tareas como completadas.

class Tarea2(val descripcion: String) {
    var completada = false

    fun marcarCompletada() {
        completada = true
    }

    override fun toString(): String {
        val estado = if (completada) "Completada" else "Pendiente"
        return "$descripcion - $estado"
    }
}

class Proyecto2(val nombre: String) {
    private val tareas = mutableListOf<Tarea2>()

    fun agregarTarea(tarea: Tarea2) {
        tareas.add(tarea)
        println("Tarea \"${tarea.descripcion}\" añadida al proyecto $nombre")
    }

    fun mostrarTareas() {
        if (tareas.isEmpty()) {
            println("No hay tareas en este proyecto.")
        } else {
            println("Tareas en el proyecto $nombre:")
            tareas.forEachIndexed { i, tarea -> println("${i + 1}. $tarea") }
        }
    }

    fun marcarTareaCompletada(indice: Int) {
        if (indice in tareas.indices) {
            tareas[indice].marcarCompletada()
            println("Tarea marcada como completada.")
        } else {
            println("Índice de tarea no válido.")
        }
    }
}

class Usuario2(val nombre: String) {
    val proyectos = mutableListOf<Proyecto2>()

    fun crearProyecto(nombreProyecto: String) {
        proyectos.add(Proyecto2(nombreProyecto))
        println("Proyecto \"$nombreProyecto\" creado.")
    }

    fun mostrarProyectos() {
        if (proyectos.isEmpty()) {
            println("No hay proyectos disponibles.")
        } else {
            println("Proyectos de $nombre:")
            proyectos.forEachIndexed { i, proyecto -> println("${i + 1}. ${proyecto.nombre}") }
        }
    }
}

fun main() {
    val usuario = Usuario2("Usuario1")

    while (true) {
        println("\nGestor de Tareas")
        println("1. Crear proyecto")
        println("2. Agregar tarea a proyecto")
        println("3. Marcar tarea como completada")
        println("4. Mostrar proyectos y tareas")
        println("5. Salir")
        print("Seleccione una opción: ")
        when (readln().toIntOrNull()) {
            1 -> {
                print("Ingrese el nombre del proyecto: ")
                val nombreProyecto = readln()
                usuario.crearProyecto(nombreProyecto)
            }
            2 -> {
                usuario.mostrarProyectos()
                print("Seleccione el número del proyecto: ")
                val indiceProyecto = readln().toIntOrNull()?.minus(1)
                val proyecto = if (indiceProyecto != null && indiceProyecto in usuario.proyectos.indices) {
                    usuario.proyectos[indiceProyecto]
                } else null

                if (proyecto != null) {
                    print("Descripción de la tarea: ")
                    val descripcionTarea = readln()
                    proyecto.agregarTarea(Tarea2(descripcionTarea))
                }
            }
            3 -> {
                usuario.mostrarProyectos()
                print("Seleccione el número del proyecto: ")
                val indiceProyecto = readln().toIntOrNull()?.minus(1)
                val proyecto = if (indiceProyecto != null && indiceProyecto in usuario.proyectos.indices) {
                    usuario.proyectos[indiceProyecto]
                } else null

                proyecto?.let {
                    it.mostrarTareas()
                    print("Seleccione el número de la tarea a completar: ")
                    val indiceTarea = readln().toIntOrNull()?.minus(1)
                    if (indiceTarea != null) it.marcarTareaCompletada(indiceTarea)
                }
            }
            4 -> usuario.proyectos.forEach { it.mostrarTareas() }
            5 -> {
                println("Saliendo...")
                break
            }
            else -> println("Opción no válida.")
        }
    }
}
